File: modules/app/agent-repl/daemon/internal/sessionlock/sessionlock.go
// Package sessionlock answers "is a shim alive for this session?" with a
// kernel-enforced lock rather than bookkeeping the daemon has to get right.
//
// # Why this exists
//
// While each shim listened on its own session-<id>.sock, uniqueness was free:
// only one process can bind a path, so a second shim for the same session was
// unreachable — bind() returned EADDRINUSE and it died. Once shims dial OUT to
// the daemon instead (design-shim-transport-inversion.md), nothing stops two
// processes claiming one session, and two shims on one conversation means two
// writers on one transcript.
//
// Connection tracking alone cannot close this. On a fresh daemon boot a
// surviving shim may not have dialled in yet, so "do I have a connection for
// this session?" answers NO when the truth is NOT YET, and the daemon would
// spawn a duplicate of a shim that is alive and mid-turn.
//
// So the shim takes TWO exclusive locks at startup — one keyed by session id,
// one keyed by workspace directory — and holds both for its lifetime. The
// session key names one conversation attempt; the workspace key names the thing
// the invariant is about, and only it catches two daemon session ids pointed at
// one workspace over one transcript (WorkspaceLockPath). Each lock is:
//
//   - kernel-enforced, so exclusion is not advisory bookkeeping;
//   - released automatically when the holder dies, however it dies, so there
//     is no stale state to reap and no PID-reuse hazard;
//   - testable synchronously, which is what covers the boot window that
//     connection tracking misses.
//
// # Platform
//
// BSD flock(2) semantics. The shim side takes the same lock from Node via
// open(2)'s O_EXLOCK, which is the same underlying lock on macOS/BSD, so the
// two interoperate. Linux has no O_EXLOCK, so the shim cannot take this lock
// there without native code — Held reports that loudly rather than pretending
// a session is free.
package sessionlock

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
	"syscall"
)

// Dir returns the directory session locks live in: a sibling of sock/ and
// store/ under the agent-repl cache root. A lock is not a socket, so it does
// not live among them — the same split as /var/run.
func Dir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("sessionlock: resolving home dir: %w", err)
	}
	return filepath.Join(home, ".cache", "agent-repl", "run"), nil
}

// Path returns the lock file for sessionID.
func Path(sessionID string) (string, error) {
	dir, err := Dir()
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, "session-"+sessionID+".lock"), nil
}

// WorkspaceKey is a workspace directory's identity in a file name: the first
// eight hex digits of the MD5 of its cleaned absolute path.
//
// It is the SAME derivation as the workspace_id on every canonical log record
// (dlog.WorkspaceFromDirectory, and the shim's own log.ts), so a lock file is
// greppable against the log lines of the shim holding it. The shim recomputes
// it in Node from the --cwd the daemon hands it — already symlink-resolved and
// absolute — and both sides normalize with their platform's path cleaner, so
// one workspace derives one path on both sides.
func WorkspaceKey(cwd string) string {
	sum := md5.Sum([]byte(filepath.Clean(cwd)))
	return hex.EncodeToString(sum[:])[:8]
}

// WorkspaceLockPath returns the lock file for the workspace rooted at cwd.
//
// It is a SECOND lock, held alongside the session lock rather than instead of
// it, because the two answer different questions. A session id names one
// daemon-side conversation attempt; a workspace names the thing the invariant
// is about — a workspace and each resumed transcript keep exactly one live
// session at a time. Two daemon session ids can point at one workspace over one
// vendor transcript, and each would take its own session lock and exclude
// nothing.
//
// The key is the CWD rather than the vendor transcript uuid because a FRESH
// session has no transcript yet, so a transcript key cannot cover the window in
// which a duplicate gets spawned.
func WorkspaceLockPath(cwd string) (string, error) {
	if cwd == "" {
		return "", fmt.Errorf("sessionlock: a workspace lock path needs a workspace directory, got an empty one")
	}
	dir, err := Dir()
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, "workspace-"+WorkspaceKey(cwd)+".lock"), nil
}

// WorkspaceLockHeld reports whether a live process holds the workspace's lock —
// i.e. whether a shim is alive for that workspace, INCLUDING one that has not
// dialled in yet. It answers exactly as Held does, and its errors mean the same
// thing: "I could not tell", which a caller must never read as free.
func WorkspaceLockHeld(cwd string) (bool, error) {
	path, err := WorkspaceLockPath(cwd)
	if err != nil {
		return false, err
	}
	return heldAt(path)
}

// EnsureDir creates the lock directory. Called at daemon boot alongside the
// other runtime dirs; the shim needs it to exist before it can take a lock.
func EnsureDir() error {
	dir, err := Dir()
	if err != nil {
		return err
	}
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("sessionlock: creating %s: %w", dir, err)
	}
	return nil
}

// Held reports whether a live process holds sessionID's lock — i.e. whether a
// shim is alive for that session, INCLUDING one that has not dialled in yet.
//
// It answers by trying to take the lock non-blockingly: success means nobody
// held it (the lock is released again immediately), failure with EWOULDBLOCK
// means someone does. A missing lock file means no shim has ever locked this
// session, which is a free session, not an error.
//
// Any other error is returned rather than guessed at: the caller must not read
// "I could not tell" as "free" and spawn a duplicate.
func Held(sessionID string) (bool, error) {
	return heldAt(pathOrEmpty(sessionID))
}

// HeldAt is Held against an explicit path (tests, and callers that already
// resolved it).
func HeldAt(path string) (bool, error) { return heldAt(path) }

func pathOrEmpty(sessionID string) string {
	p, err := Path(sessionID)
	if err != nil {
		return ""
	}
	return p
}

func heldAt(path string) (bool, error) {
	if path == "" {
		return false, fmt.Errorf("sessionlock: empty lock path")
	}
	f, err := os.OpenFile(path, os.O_RDWR, 0o644)
	if err != nil {
		if os.IsNotExist(err) {
			// No lock file: no shim has ever claimed this session.
			return false, nil
		}
		return false, fmt.Errorf("sessionlock: opening %s: %w", path, err)
	}
	defer f.Close()

	err = syscall.Flock(int(f.Fd()), syscall.LOCK_EX|syscall.LOCK_NB)
	if err == nil {
		// We got it, so nobody held it. Release immediately — this is a probe,
		// not a claim; the daemon never holds a session lock (it must survive
		// the daemon, not track it).
		_ = syscall.Flock(int(f.Fd()), syscall.LOCK_UN)
		return false, nil
	}
	if err == syscall.EWOULDBLOCK {
		return true, nil
	}
	return false, fmt.Errorf("sessionlock: probing %s: %w", path, err)
}

File: modules/app/agent-repl/agent-shim/claude/shim/src/uds/session-lock.ts
/**
 * The shim's exclusive claim on its session.
 *
 * # Why
 *
 * While each shim LISTENED on its own `session-<id>.sock`, uniqueness was free:
 * only one process can bind a path, so a second shim for the same session was
 * unreachable — bind(2) returned EADDRINUSE and it died. Now that shims dial
 * OUT to the daemon (design-shim-transport-inversion.md) nothing stops two
 * processes claiming one session, and two shims on one conversation means two
 * writers on one transcript.
 *
 * The daemon cannot close that gap by tracking connections alone. On a fresh
 * boot a surviving shim may not have dialled in yet, so "do I have a connection
 * for this session?" answers NO when the truth is NOT YET, and the daemon would
 * spawn a duplicate of a shim that is alive and mid-turn.
 *
 * So the shim takes a kernel-enforced lock at startup and holds it for its
 * lifetime. Held is what the daemon probes before spawning.
 *
 * # Two keys, both required
 *
 * A session id names one daemon-side conversation attempt; a WORKSPACE names
 * the thing the invariant is actually about. A workspace and each resumed
 * transcript keep exactly one live session at a time, and two daemon session
 * ids can point at one workspace and one vendor transcript — so a claim keyed
 * only by session id lets two shims take two different locks over one
 * transcript and excludes nothing.
 *
 * The workspace key is the CWD rather than the vendor transcript uuid because a
 * FRESH session has no transcript yet: a transcript key cannot cover the window
 * in which the duplicate is spawned.
 *
 * Both locks are taken, in a fixed order — session lock first, then workspace
 * lock — so no two shims can ever take them in opposite orders. Failing to take
 * either is a refusal to start.
 *
 * # Mechanism
 *
 * `open(2)`'s `O_EXLOCK`, which takes a BSD flock as part of opening the file —
 * the same lock Go's `syscall.Flock` takes, so the daemon's probe and this
 * claim interoperate. Two properties matter and neither is available from a
 * plain lock FILE:
 *
 *   - the kernel enforces it, so exclusion is not advisory bookkeeping; and
 *   - it is released automatically when the process dies, however it dies, so
 *     there is no stale lock to reap and no PID-reuse hazard.
 *
 * `O_EXLOCK` is macOS/BSD only. Linux has no equivalent in Node without native
 * code, so there acquiring FAILS LOUDLY rather than pretending the session is
 * claimed — a silent no-op would hand the daemon a false "free" and let it
 * spawn the duplicate this exists to prevent.
 */
import fs from "node:fs";
import path from "node:path";
import os from "node:os";
import { createHash } from "node:crypto";
import { bindLog } from "./log.js";

const COMPONENT = "shim-session-lock";
const LOGGER = bindLog({ component: COMPONENT, operation: "shim.session-lock.lifecycle" });

/**
 * `O_EXLOCK` from `<sys/fcntl.h>`: take an exclusive advisory lock as part of
 * open(2), the same lock `flock(2)` takes — so the daemon's Go-side probe
 * (`syscall.Flock`) and this claim contend correctly with each other. Verified
 * both directions: while Node holds it Go gets EWOULDBLOCK, and once Node
 * closes the fd Go acquires.
 *
 * Spelled numerically because Node does not expose it on `fs.constants`, even
 * on platforms whose open(2) honors it — flags are passed straight through, so
 * the raw value works. BSD/macOS only; undefined elsewhere, which
 * acquireSessionLock turns into a loud refusal rather than a silent no-op.
 */
const O_EXLOCK: number | undefined =
  process.platform === "darwin" || process.platform.endsWith("bsd") ? 0x20 : undefined;

/** The directory session locks live in: a sibling of sock/ and store/. */
export function lockDir(): string {
  return path.join(os.homedir(), ".cache", "agent-repl", "run");
}

/** The lock file for sessionId. */
export function lockPath(sessionId: string): string {
  return path.join(lockDir(), `session-${sessionId}.lock`);
}

/**
 * The workspace's identity in a file name: the first eight hex digits of the
 * MD5 of its normalized absolute directory.
 *
 * This is the SAME derivation as the `workspace_id` on every canonical log
 * record (log.ts, and dlog.WorkspaceFromDirectory on the Go side), so a lock
 * file is greppable against the log lines of the shim holding it. The daemon
 * hands the shim an already symlink-resolved absolute `--cwd`, and both sides
 * normalize identically (cleanForKey is Go's filepath.Clean), so the Node and
 * Go derivations agree on one path for one workspace.
 */
export function workspaceLockKey(cwd: string): string {
  if (cwd === "") {
    throw new Error(`${COMPONENT}: a workspace lock needs an absolute workspace directory, got ""`);
  }
  return createHash("md5").update(cleanForKey(cwd)).digest("hex").slice(0, 8);
}

/**
 * Go's `filepath.Clean`, spelled in Node. `path.normalize` collapses `.`, `..`
 * and repeated separators the same way and differs in exactly one respect: it
 * KEEPS a trailing separator where Clean strips it. Unreconciled, that one
 * difference gives `/ws` and `/ws/` two locks on the Node side and one on the
 * Go side — two shims over one workspace, which is what the lock exists to stop.
 */
function cleanForKey(p: string): string {
  const normalized = path.normalize(p);
  return normalized.length > 1 ? normalized.replace(/\/+$/, "") : normalized;
}

/** The lock file for the workspace rooted at cwd. */
export function workspaceLockPath(cwd: string): string {
  return path.join(lockDir(), `workspace-${workspaceLockKey(cwd)}.lock`);
}

/**
 * Take this session's exclusive lock and hold it until the process exits.
 *
 * Returns a release function for a deliberate teardown. Not calling it is fine
 * and is the normal path — the kernel drops the lock when the process dies,
 * which is precisely the property that makes the lock trustworthy.
 *
 * Throws when the lock is already held (another shim owns this session) or when
 * the platform cannot take it. Both are refusals to run, not warnings: a shim
 * that cannot prove it is the only one for its session must not start.
 */
export function acquireSessionLock(sessionId: string): () => void {
  return acquireExclusiveLock({
    kind: "session",
    subject: `session ${sessionId}`,
    file: lockPath(sessionId),
    context: { agent_repl_session_id: sessionId },
  });
}

/**
 * Take the workspace's exclusive lock and hold it until the process exits.
 *
 * Identical in contract to acquireSessionLock, and taken AFTER it: the fixed
 * order is what keeps two shims racing for the same pair from deadlocking each
 * other. Throws when another shim already owns the workspace or when the
 * platform cannot take the lock; a shim that cannot prove it is the workspace's
 * only one must not start.
 */
export function acquireWorkspaceLock(cwd: string): () => void {
  return acquireExclusiveLock({
    kind: "workspace",
    subject: `workspace ${cwd}`,
    file: workspaceLockPath(cwd),
    context: { workspace_dir: cwd, workspace_id: workspaceLockKey(cwd) },
  });
}

/** One kernel-enforced claim, which is the whole of both locks' mechanism. */
function acquireExclusiveLock(claim: {
  kind: string;
  subject: string;
  file: string;
  context: Record<string, unknown>;
}): () => void {
  LOGGER.log({ ...claim.context, platform: process.platform }, `acquiring exclusive shim ${claim.kind} lock`);
  if (O_EXLOCK === undefined) {
    throw new Error(
      `${COMPONENT}: ${process.platform} has no O_EXLOCK, so the shim cannot claim ${claim.subject} ` +
        `exclusively; refusing to start rather than risk two shims writing one transcript`,
    );
  }
  const file = claim.file;
  fs.mkdirSync(path.dirname(file), { recursive: true });

  let fd: number;
  try {
    // O_NONBLOCK so an already-held lock fails immediately instead of hanging
    // this process behind whichever shim owns the claim.
    fd = fs.openSync(file, fs.constants.O_CREAT | fs.constants.O_RDWR | O_EXLOCK | fs.constants.O_NONBLOCK);
  } catch (err) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code === "EAGAIN" || code === "EWOULDBLOCK") {
      throw new Error(
        `${COMPONENT}: ${claim.subject} is already held by another shim (${file}); refusing to start a duplicate`,
      );
    }
    throw new Error(`${COMPONENT}: cannot take the ${claim.kind} lock ${file}: ${(err as Error).message}`);
  }

  LOGGER.log(claim.context, `holding ${claim.kind} lock ${file}`);
  let released = false;
  return () => {
    if (released) return;
    released = true;
    try {
      fs.closeSync(fd); // closing drops the flock
      LOGGER.log({ ...claim.context, lock_path: file }, `released exclusive shim ${claim.kind} lock`);
    } catch (err) {
      LOGGER.log({ level: "error", ...claim.context, cause: err }, `releasing the ${claim.kind} lock failed: ${(err as Error).message}`);
    }
  };
}
